code/modules/awaymissions/mission_code/mining.dm

```diff
@@ -0,0 +1,398 @@
+	name = "beacon beacon"
+	desc = "Creates holographic versions of a destructive magical club."
+	icon_state = "hierophant"
+	mob_types = list(/mob/living/simple_animal/hostile/megafauna/hierophant/virtual)
+
+/obj/structure/spawner/megafauna/legion
+	name = "skull beacon"
+	desc = "Creates holographic versions of a gigantic skull demon guarding the necropolis."
+	icon_state = "legion"
+	mob_types = list(/mob/living/simple_animal/hostile/megafauna/legion/virtual)
+
+//							 //
+// ----- Recall Portals ---- //
+//							 //
+
+/obj/effect/proc_holder/spell/portal_recall
+	name = "Portal Recall"
+	desc = "This will teleport you back to your previously used portal. One use only."
+	clothes_req = FALSE
+	action_icon_state = "blink"
+	var/list/recall_portals = list()
+
+/obj/effect/proc_holder/spell/portal_recall/Click(mob/user = usr)
+	if(!recall_portals.len)
+		user.mind.RemoveSpell(src) // remove spell if no portals left
+	var/turf/recall_turf = get_turf(recall_portals[recall_portals.len])
+	if(recall_turf)
+		do_teleport(user, recall_turf, 0, no_effects = FALSE, channel = TELEPORT_CHANNEL_BLUESPACE)
+		recall_portals -= recall_portals[recall_portals.len]
+		if(!recall_portals.len)
+			user.mind.RemoveSpell(src) // remove spell if no portals left
+
+/obj/effect/portal/permanent/one_way/recall
+	name = "recall portal"
+	desc = "Gives you a one time ability to return to this portal once you have entered."
+	mech_sized = TRUE
+	keep = TRUE
+
+/obj/effect/portal/permanent/one_way/recall/Crossed(atom/movable/AM, oldloc)
+	if(ismob(AM))
+		var/mob/user = AM
+		var/check = locate(/obj/effect/proc_holder/spell/portal_recall) in user.mind.spell_list
+		if(check)
+			var/obj/effect/proc_holder/spell/portal_recall/mob_recall = check
+			for(var/obj/effect/portal/permanent/one_way/recall/P in mob_recall.recall_portals)
+				if(src == P)
+					return ..(AM, oldloc, force_stop = 1) // don't teleport if they have a recall spell with this portal already (or have just teleported onto it)
+	return ..()
+
+/obj/effect/portal/permanent/one_way/recall/teleport(atom/movable/M, force = FALSE)
+	. = ..()
+	if(ismob(M))
+		var/mob/user = M
+		var/findspell = locate(/obj/effect/proc_holder/spell/portal_recall) in user.mind.spell_list
+		var/obj/effect/proc_holder/spell/portal_recall/personal_recall = findspell ? findspell : new
+		personal_recall.recall_portals += src
+		if(!findspell)
+			user.mind.AddSpell(personal_recall)
+
+//							   //
+// -----Virtual Megafauna----- //
+//							   //
+
+/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner/virtual
+	name = "blood-drunk miner hologram"
+	desc = "A holographic miner, eternally hunting."
+	crusher_loot = list()
+	loot = list()
+	virtual = 1
+
+/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner/virtual/Initialize()
+	. = ..()
+	qdel(internal)
+
+/mob/living/simple_animal/hostile/megafauna/blood_drunk_miner/virtual/grant_achievement(medaltype, scoretype, crusher_kill)
```

Comment:
you can edit the base proc to account for virtuals since you used it as a base var.